### PR TITLE
fix: wgrad is invalid when enable DDP on mcore

### PR DESCRIPTION
**Pull request by RuibinCheung**
# Description

 When `overlap_grad_reduce` on mcore is True, need to ensure that backward hooks are all run on the main backprop thread to prevent deadlocks. We need to setup wgrad tensor to prevent backward hooks from being run in a background thread.
 
 NV upstream code: https://github.com/NVIDIA/TransformerEngine/blob/c55e425ac5751c72bfa656560fc7c65ab7d895e9/transformer_engine/pytorch/module/linear.py#L697-L703

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/tensor/quantized_tensor.py [hunk 1]
```diff
@@ -27,11 +27,8 @@ def prepare_for_saving(
     # pylint: disable=unidiomatic-typecheck  # Using type instead of isinstance to check exact type
     tensor_list, tensor_objects_list = [], []
     for tensor in tensors:
-        if tensor is None:
```
Comment (RuibinCheung) [inline transformer_engine/pytorch/tensor/quantized_tensor.py:30]: Current code will lose parameters' attributes when saving `torch.nn.Parameter`. We need to keep `grad_added_to_main_grad` attribute to make gradient accumulate fusion work. I think this is a NV TE's bug on v2.1 and fixed it on v2.2.

This change is from NV v2.2 upstream code: https://github.com/NVIDIA/TransformerEngine/blob/c55e425ac5751c72bfa656560fc7c65ab7d895e9/transformer_engine/pytorch/tensor/quantized_tensor.py#L18-L34

**Comment by wenchenvincent** [comment]
This fix came in when IFU v2.2 merged into dev and it should also be available in release v2.1. Thus closing.
